Source/dvlnet/packet.h

```diff
@@ -150,67 +153,68 @@ class packet_out : public packet_proc<packet_out> {
 };
 
 template <class P>
-void packet_proc<P>::process_data()
+tl::expected<void, PacketError> packet_proc<P>::process_data()
 {
 	P &self = static_cast<P &>(*this);
-	self.process_element(m_type);
-	self.process_element(m_src);
-	self.process_element(m_dest);
+	{
+		tl::expected<void, PacketError> result
+		    = self.process_element(m_type)
+		          .and_then([&]() {
+			          return self.process_element(m_src);
+		          })
+		          .and_then([&]() {
+			          return self.process_element(m_dest);
+		          });
+		if (!result.has_value())
+			return result;
+	}
 	switch (m_type) {
 	case PT_MESSAGE:
-		self.process_element(m_message);
-		break;
+		return self.process_element(m_message);
 	case PT_TURN:
-		self.process_element(m_turn.SequenceNumber);
-		self.process_element(m_turn.Value);
-		break;
+		return self.process_element(m_turn.SequenceNumber)
+		    .and_then([&]() { return self.process_element(m_turn.Value); });
 	case PT_JOIN_REQUEST:
-		self.process_element(m_cookie);
-		self.process_element(m_info);
-		break;
+		return self.process_element(m_cookie)
+		    .and_then([&]() { return self.process_element(m_info); });
 	case PT_JOIN_ACCEPT:
-		self.process_element(m_cookie);
-		self.process_element(m_newplr);
-		self.process_element(m_info);
-		break;
+		return self.process_element(m_cookie)
+		    .and_then([&]() { return self.process_element(m_newplr); })
+		    .and_then([&]() { return self.process_element(m_info); });
 	case PT_CONNECT:
-		self.process_element(m_newplr);
-		self.process_element(m_info);
-		break;
+		return self.process_element(m_newplr)
+		    .and_then([&]() { return self.process_element(m_info); });
 	case PT_DISCONNECT:
-		self.process_element(m_newplr);
-		self.process_element(m_leaveinfo);
-		break;
+		return self.process_element(m_newplr)
+		    .and_then([&]() { return self.process_element(m_leaveinfo); });
 	case PT_INFO_REPLY:
-		self.process_element(m_info);
-		break;
+		return self.process_element(m_info);
 	case PT_INFO_REQUEST:
-		break;
+		return {};
 	case PT_ECHO_REQUEST:
 	case PT_ECHO_REPLY:
-		self.process_element(m_time);
-		break;
+		return self.process_element(m_time);
 	}
+	app_fatal(StrCat("Unkown packet type: ", static_cast<int>(m_type)));

```

Comment:
`packet_in::process_element()` will read `m_type` from the buffer received from the remote client. I don't think it's a good idea to crash the game client if the packet type we receive is invalid.

Reply:
Done, we now actually return an `unexpected` instead of silently ignoring such packets.